pytensor/tensor/slinalg.py

```diff
@@ -1703,6 +1706,402 @@ def block_diag(*matrices: TensorVariable):
     return _block_diagonal_matrix(*matrices)
 
 
+class QR(Op):
+    """
+    QR Decomposition
+    """
+
+    __props__ = (
+        "overwrite_a",
+        "mode",
+        "lwork_cache",
+        "pivoting",
+        "check_finite",
+        "last_shape_cache",
+    )
+
+    def __init__(
+        self,
+        overwrite_a: bool,
+        mode: Literal["full", "r", "economic", "raw"] = "full",
+        lwork_cache: tuple[int, int] = (-1, -1),
+        pivoting: bool = False,
+        check_finite: bool = False,
+        last_shape_cache: tuple[tuple[int, int], tuple[int, int]] = ((0, 0), (0, 0)),
+    ):
+        self.mode = mode
+        self.overwrite_a = overwrite_a
+        self.lwork_cache = lwork_cache
+        self.pivoting = pivoting
+        self.check_finite = check_finite
+
+        self.destroy_map = {}
+
+        if overwrite_a:
+            self.destroy_map = {0: [0]}
+
+        match self.mode:
+            case "economic":
+                self.gufunc_signature = "(m,n)->(m,k),(k,n)"
+            case "full":
+                self.gufunc_signature = "(m,n)->(m,m),(m,n)"
+            case "r":
+                self.gufunc_signature = "(m,n)->(m,n)"
+            case "raw":
+                self.gufunc_signature = "(m,n)->(n,m),(k),(m,n)"
+            case _:
+                raise ValueError(
+                    f"Invalid mode '{mode}'. Supported modes are 'full', 'economic', 'r', and 'raw'."
+                )
+
+        if pivoting:
+            self.gufunc_signature += ",(n)"
+
+        # Used to cache lwork size in the C backend, where possible
+        # The idea here is that in vectorized cases (e.g. blockwise), the input shape will not change between perform
+        # calls, so we can optimize the LAPACK call to avoid recomputing the optimal lwork size every time.
+        self.last_shape_cache = last_shape_cache
+
+    def make_node(self, x):
+        x = as_tensor_variable(x)
+
+        assert x.ndim == 2, "The input of qr function should be a matrix."
+
+        # Preserve static shape information if possible
+        M, N = x.type.shape
+        if M is not None and N is not None:
+            K = min(M, N)
+        else:
+            K = None
+
+        in_dtype = x.type.numpy_dtype
+        out_dtype = np.dtype(f"f{in_dtype.itemsize}")
+
+        match self.mode:
+            case "full":
+                outputs = [
+                    tensor("Q", shape=(M, M), dtype=out_dtype),
```

Comment:
It's better not to give names to intermediate variables, this would shadow a user `Q`, if they wanted to do `eval({"Q": something})` or find a variable by name